src/cli.ts

```diff
@@ -283,33 +286,53 @@ yargs.command(
         try {
             const adapter = createAdapterFromUrl(argv.db!)
             await adapter.init()
-            const head = await getHead()
-            const status = await getStatus(adapter, head)
-            process.stdout.write(status.toString())
-            if (status.newCommits.some(commit => commit.tasks.length > 0)) {
-                if (argv.confirm) {
-                    const answer = await inquirer.prompt<{ continue: boolean }>({
-                        type: 'confirm',
-                        name: 'continue',
-                        message: 'Continue?',
-                    })
-                    if (!answer.continue) {
-                        process.exit(0)
+            while (true) {
+                const head = await getHead()
+                const status = await getStatus(adapter, head)
+                process.stdout.write(status.toString())
+                const tasks = status.newCommits.reduce<Task[]>((prev, next) => prev.concat(next.tasks), [])
+                if (tasks.length > 0) {
+                    if (argv.confirm) {
+                        const answer = await inquirer.prompt<{ continue: boolean }>({
+                            type: 'confirm',
+                            name: 'continue',
+                            message: 'Continue?',
+                        })
+                        if (!answer.continue) {
+                            process.exit(0)
+                        }
+                        process.stdout.write('\n')
                     }
-                    process.stdout.write('\n')
-                }
-                process.stdout.write('Starting migration\n\n')
-                for (const commit of status.newCommits) {
-                    process.stdout.write(`${chalk.yellow(commit.shortSha1)} ${commit.subject}\n`)
-                    for (const task of commit.tasks) {
-                        process.stdout.write(task.toString() + ' ...')
-                        const interval = setInterval(() => process.stdout.write('.'), 100)
-                        await task.execute(argv.migrationOutDir!, adapter, head, commit)
-                        clearInterval(interval)
-                        process.stdout.write(' Success\n')
+
+                    process.stdout.write('Starting migration\n\n')
+
+                    const hasChanged = await adapter.waitForPending(CLI_LOGGER)
+
+                    if (hasChanged) {
+                        process.stdout.write('The migrations have changed, reloading..\n\n')
+                        continue
+                    }
+                    // create pending tasks
+                    for (const task of tasks) {
+                        await adapter.beginMigrationTask(task)
```

Comment:
Between this and `waitForPendingMigrations()`, could another process have started the same migration?

Reply:
hm yes, I added some failure recovery